### PR TITLE
Added recourse_methods/face_method.py

### PR DESCRIPTION
**Pull request by jakeval**
# Changes Made
* Added face_method.py
* Added test_face_method.py

# Feedback Requested
* is face_method.py clear and maintainable?
* is test_face_method.py clear? Are the comments sufficient?

### CODE REVIEW

File: recourse_methods/test_face_method.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        "recourse_methods.face_method.recourse_adapter.RecourseAdapter",
+        autospec=True,
+    )
+    @mock.patch("recourse_methods.face_method._get_edge_weight")
+    def test_append_new_point(self, mock_get_edge_weight, mock_adapter):
+        distance_threshold = 1
+        dataset = pd.DataFrame({"a": [0, 1, 2], "label": [0, 1, 1]})
+
+        # The graph to append a new point to
+        graph = sparse.csr_array(
```
Comment (bringingjoy) [inline recourse_methods/test_face_method.py:153]: Can you please simplify this test example? The initial graph and clarify the input dataset "a" - is that a dataset with a single 3-dimensional point, or three 1-dimensional point? It's hard to discern from the setup.

**Review by bringingjoy — CHANGES_REQUESTED** [review]

File: recourse_methods/test_face_method.py (hunk 1 above)
Comment (bringingjoy) [reply recourse_methods/test_face_method.py:153]: From offline discussion with @jakeval , this test function currently
* calculates the distance from the new point to every other point
* calculates the weight from the distance
* replaces over-threshold distances with 0-weight
* adds the distance vector to the original graph as a column and a row

Would you please split these out into individual test functions? I think that'll simplify this and make this easier to maintain. Thank you!

File: recourse_methods/face_method.py [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+import numpy as np
+import numba
+from scipy import sparse
+from data import recourse_adapter
+from recourse_methods import base_type
+from models import model_interface
+
+
+@numba.jit(nopython=True)
+def _get_edge_weight(distance, weight_bias):
```
[the remark this answers is not in the document]
Comment (jakeval) [reply recourse_methods/face_method.py:13]: This is a [limitation of numba](https://github.com/numba/numba/issues/5409) -- it has trouble with class members. There may be a cleaner way of handling this, but I didn't want to spend too much time when this seems like an alright solution.

File: recourse_methods/face_method.py [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+            weight_bias=self.weight_bias,
+        )
+        sparse_graph_weights = sparse.csr_array(graph_weights)
+        if filepath_to_save_to is not None:
+            sparse.save_npz(filepath_to_save_to, sparse_graph_weights)
+        self.graph = sparse_graph_weights
+
+    @staticmethod
+    @numba.jit(nopython=True)
+    def _get_e_graph_weights(
```
[the remark this answers is not in the document]
Comment (jakeval) [reply recourse_methods/face_method.py:117]: This is a good idea. The current order was just how it ended up organically. I've reordered it to be more readable.

File: recourse_methods/test_face_method.py (hunk 1 above)
Comment (jakeval) [reply recourse_methods/test_face_method.py:153]: The function is now called _add_point_to_graph() and is no longer directly tested. Instead, it uses the helper functions _calculate_weight_vector() and _add_edges_to_graph() which are unit tested.

**Review by bringingjoy — APPROVED** [review]
The testing is a lot clearer - thanks!

**Pull request merged**